File: cpp/dev/src/bag.cpp
// bag.cpp -- ROS bag v2.0 reader/writer implementation.
#include "irap_noroslib/bag.hpp"

#include <cstring>
#include <stdexcept>

namespace irap_noroslib {
namespace {

const size_t kChunkThreshold = 768 * 1024;   // rosbag's default chunk size

// -- little-endian pack helpers (bag integers are always LE) -----------------
void put_u8(std::string* s, uint8_t v) { s->push_back((char)v); }
void put_u32(std::string* s, uint32_t v) {
  for (int i = 0; i < 4; ++i) s->push_back((char)((v >> (8 * i)) & 0xff));
}
void put_u64(std::string* s, uint64_t v) {
  for (int i = 0; i < 8; ++i) s->push_back((char)((v >> (8 * i)) & 0xff));
}
std::string u8(uint8_t v) { std::string s; put_u8(&s, v); return s; }
std::string u32(uint32_t v) { std::string s; put_u32(&s, v); return s; }
std::string u64(uint64_t v) { std::string s; put_u64(&s, v); return s; }
std::string tm(uint32_t s_, uint32_t n_) { std::string s; put_u32(&s, s_); put_u32(&s, n_); return s; }

uint32_t get_u32(const std::string& s, size_t off) {
  uint32_t v = 0;
  for (int i = 0; i < 4; ++i) v |= (uint32_t)(uint8_t)s[off + i] << (8 * i);
  return v;
}
uint64_t get_u64(const std::string& s, size_t off) {
  uint64_t v = 0;
  for (int i = 0; i < 8; ++i) v |= (uint64_t)(uint8_t)s[off + i] << (8 * i);
  return v;
}

// A single header field: <4-byte len><name=value>.
std::string enc_field(const std::string& name, const std::string& value) {
  std::string field = name + "=" + value;
  return u32((uint32_t)field.size()) + field;
}

// The raw field sequence (no outer length) -- a connection record's DATA.
std::string encode_header_fields(
    const std::vector<std::pair<std::string, std::string>>& fields) {
  std::string out;
  for (const auto& kv : fields) out += enc_field(kv.first, kv.second);
  return out;
}

// A length-prefixed header block: <4-byte len><fields> -- a record HEADER.
std::string encode_header(
    const std::vector<std::pair<std::string, std::string>>& fields) {
  std::string body = encode_header_fields(fields);
  return u32((uint32_t)body.size()) + body;
}

std::map<std::string, std::string> decode_header_fields(const std::string& blob) {
  std::map<std::string, std::string> out;
  size_t i = 0;
  while (i + 4 <= blob.size()) {
    uint32_t flen = get_u32(blob, i);
    i += 4;
    if (i + flen > blob.size()) break;
    std::string field = blob.substr(i, flen);
    i += flen;
    size_t eq = field.find('=');
    if (eq != std::string::npos) out[field.substr(0, eq)] = field.substr(eq + 1);
  }
  return out;
}

std::string conn_header_data(const BagConnection& c) {
  std::vector<std::pair<std::string, std::string>> f = {
      {"topic", c.topic}, {"type", c.type}, {"md5sum", c.md5},
      {"message_definition", c.definition}};
  if (!c.callerid.empty()) f.push_back({"callerid", c.callerid});
  return encode_header_fields(f);
}

std::string record_bytes(const std::vector<std::pair<std::string, std::string>>& fields,
                         const std::string& data) {
  return encode_header(fields) + u32((uint32_t)data.size()) + data;
}

}  // namespace

// ---------------------------------------------------------------- writer ----
BagWriter::BagWriter(const std::string& path) : path_(path) {
  f_.open(path, std::ios::binary | std::ios::out | std::ios::trunc);
  if (!f_) throw std::runtime_error("nr_rosbag: cannot open " + path + " for writing");
  f_.write("#ROSBAG V2.0\n", 13);
  bagheader_pos_ = f_.tellp();
  std::string hdr = bag_header_record(0, 0, 0);
  f_.write(hdr.data(), (std::streamsize)hdr.size());
}

BagWriter::~BagWriter() { close(); }

int BagWriter::connection(const std::string& topic, const std::string& type,
                          const std::string& md5, const std::string& definition,
                          const std::string& callerid) {
  auto it = conns_.find(topic);
  if (it != conns_.end()) return it->second.id;
  BagConnection c;
  c.id = (int)conns_.size();
  c.topic = topic; c.type = type; c.md5 = md5;
  c.definition = definition; c.callerid = callerid;
  auto& stored = conns_[topic] = c;
  by_id_[stored.id] = &stored;
  return stored.id;
}

void BagWriter::write(int conn, uint32_t secs, uint32_t nsecs,
                      const std::vector<uint8_t>& body) {
  chunk_.push_back({conn, secs, nsecs, body});
  chunk_bytes_ += body.size();
  if (by_id_.count(conn)) by_id_[conn]->count++;
  if (chunk_bytes_ >= kChunkThreshold) flush_chunk();
}

void BagWriter::write_record(
    const std::vector<std::pair<std::string, std::string>>& fields,
    const std::string& data) {
  std::string rec = record_bytes(fields, data);
  f_.write(rec.data(), (std::streamsize)rec.size());
}

std::string BagWriter::bag_header_record(uint64_t index_pos, uint32_t conn_count,
                                         uint32_t chunk_count) {
  std::string header = encode_header({
      {"op", u8(BAG_OP_BAG_HEADER)}, {"index_pos", u64(index_pos)},
      {"conn_count", u32(conn_count)}, {"chunk_count", u32(chunk_count)}});
  int pad = 4096 - (int)header.size() - 4;
  if (pad < 0) pad = 0;
  return header + u32((uint32_t)pad) + std::string(pad, ' ');
}

void BagWriter::flush_chunk() {
  if (chunk_.empty()) return;
  uint64_t chunk_pos = (uint64_t)f_.tellp();
  std::string data;
  std::map<int, bool> seen;
  std::map<int, std::vector<std::pair<std::pair<uint32_t, uint32_t>, uint32_t>>> index;
  std::map<int, uint32_t> counts;
  uint32_t ss = 0, sn = 0, es = 0, en = 0;
  bool first = true;
  for (const auto& m : chunk_) {
    if (!seen[m.conn]) {
      seen[m.conn] = true;
      const BagConnection& c = *by_id_[m.conn];
      data += record_bytes({{"op", u8(BAG_OP_CONNECTION)}, {"conn", u32(m.conn)},
                            {"topic", c.topic}}, conn_header_data(c));
    }
    uint32_t offset = (uint32_t)data.size();
    std::string body((const char*)m.body.data(), m.body.size());
    data += record_bytes({{"op", u8(BAG_OP_MSG_DATA)}, {"conn", u32(m.conn)},
                          {"time", tm(m.secs, m.nsecs)}}, body);
    index[m.conn].push_back({{m.secs, m.nsecs}, offset});
    counts[m.conn]++;
    if (first || m.secs < ss || (m.secs == ss && m.nsecs < sn)) { ss = m.secs; sn = m.nsecs; }
    if (first || m.secs > es || (m.secs == es && m.nsecs > en)) { es = m.secs; en = m.nsecs; }
    first = false;
  }
  write_record({{"op", u8(BAG_OP_CHUNK)}, {"compression", "none"},
                {"size", u32((uint32_t)data.size())}}, data);
  for (const auto& kv : index) {
    std::string idx;
    for (const auto& e : kv.second) idx += tm(e.first.first, e.first.second) + u32(e.second);
    write_record({{"op", u8(BAG_OP_IDX_DATA)}, {"ver", u32(1)},
                  {"conn", u32(kv.first)}, {"count", u32((uint32_t)kv.second.size())}}, idx);
  }
  chunk_infos_.push_back({chunk_pos, ss, sn, es, en, counts});
  chunk_.clear();
  chunk_bytes_ = 0;
}

void BagWriter::close() {
  if (closed_) return;
  closed_ = true;
  flush_chunk();
  uint64_t index_pos = (uint64_t)f_.tellp();
  for (const auto& kv : conns_) {
    const BagConnection& c = kv.second;
    write_record({{"op", u8(BAG_OP_CONNECTION)}, {"conn", u32(c.id)},
                  {"topic", c.topic}}, conn_header_data(c));
  }
  for (const auto& ci : chunk_infos_) {
    std::string data;
    for (const auto& cc : ci.counts) { data += u32(cc.first); data += u32(cc.second); }
    write_record({{"op", u8(BAG_OP_CHUNK_INFO)}, {"ver", u32(1)},
                  {"chunk_pos", u64(ci.pos)}, {"start_time", tm(ci.ss, ci.sn)},
                  {"end_time", tm(ci.es, ci.en)},
                  {"count", u32((uint32_t)ci.counts.size())}}, data);
  }
  f_.seekp(bagheader_pos_);
  std::string hdr = bag_header_record(index_pos, (uint32_t)conns_.size(),
                                      (uint32_t)chunk_infos_.size());
  f_.write(hdr.data(), (std::streamsize)hdr.size());
  f_.close();
}

// ---------------------------------------------------------------- reader ----
BagReader::BagReader(const std::string& path) : path_(path) {
  f_.open(path, std::ios::binary | std::ios::in);
  if (!f_) throw std::runtime_error("nr_rosbag: cannot open " + path);
  std::string line;
  std::getline(f_, line);
  if (line.rfind("#ROSBAG V2.0", 0) != 0)
    throw std::runtime_error(path + " is not a ROS bag v2.0 file");
  std::streampos pos = f_.tellg();
  std::map<std::string, std::string> header;
  std::string data;
  read_record(&header, &data);
  conn_count_ = get_u32(header["conn_count"], 0);
  chunk_count_ = get_u32(header["chunk_count"], 0);
  index_pos_ = get_u64(header["index_pos"], 0);
  data_start_ = f_.tellg();
  f_.seekg(pos);
}

bool BagReader::read_record(std::map<std::string, std::string>* header, std::string* data) {
  char lenbuf[4];
  f_.read(lenbuf, 4);
  if (f_.gcount() < 4) return false;
  uint32_t hlen = 0;
  for (int i = 0; i < 4; ++i) hlen |= (uint32_t)(uint8_t)lenbuf[i] << (8 * i);
  std::string hbytes(hlen, '\0');
  f_.read(&hbytes[0], hlen);
  *header = decode_header_fields(hbytes);
  f_.read(lenbuf, 4);
  if (f_.gcount() < 4) return false;
  uint32_t dlen = 0;
  for (int i = 0; i < 4; ++i) dlen |= (uint32_t)(uint8_t)lenbuf[i] << (8 * i);
  data->assign(dlen, '\0');
  f_.read(&(*data)[0], dlen);
  return true;
}

void BagReader::parse_connection(const std::map<std::string, std::string>& header,
                                 const std::string& data) {
  int cid = (int)get_u32(header.at("conn"), 0);
  std::map<std::string, std::string> f = decode_header_fields(data);
  BagConnection c;
  c.id = cid;
  auto it = header.find("topic");
  c.topic = it != header.end() ? it->second : f["topic"];
  c.type = f["type"];
  c.md5 = f["md5sum"];
  c.definition = f["message_definition"];
  c.callerid = f["callerid"];
  conns_[cid] = c;
}

BagMessage BagReader::msg_from(const std::map<std::string, std::string>& header,
                               const std::string& data) {
  BagMessage m;
  m.conn = (int)get_u32(header.at("conn"), 0);
  const std::string& t = header.at("time");
  m.secs = get_u32(t, 0);
  m.nsecs = get_u32(t, 4);
  auto it = conns_.find(m.conn);
  m.topic = it != conns_.end() ? it->second.topic : "?";
  m.data.assign(data.begin(), data.end());
  return m;
}

void BagReader::read_chunk(
    const std::map<std::string, std::string>& header, const std::string& data_in,
    const std::function<void(const BagMessage&, const BagConnection&)>& cb) {
  auto it = header.find("compression");
  std::string comp = it != header.end() ? it->second : "none";
  if (comp != "none")
    throw std::runtime_error("nr_rosbag: this build reads only uncompressed bags "
                             "(chunk compression=" + comp + ")");
  const std::string& data = data_in;
  size_t i = 0;
  while (i + 4 <= data.size()) {
    uint32_t hlen = get_u32(data, i); i += 4;
    std::map<std::string, std::string> rhdr = decode_header_fields(data.substr(i, hlen));
    i += hlen;
    uint32_t dlen = get_u32(data, i); i += 4;
    std::string rdata = data.substr(i, dlen);
    i += dlen;
    uint8_t op = (uint8_t)rhdr["op"][0];
    if (op == BAG_OP_CONNECTION) {
      parse_connection(rhdr, rdata);
    } else if (op == BAG_OP_MSG_DATA) {
      BagMessage m = msg_from(rhdr, rdata);
      cb(m, conns_[m.conn]);
    }
  }
}

void BagReader::read_messages(
    const std::function<void(const BagMessage&, const BagConnection&)>& cb) {
  f_.clear();
  f_.seekg(data_start_);
  while (true) {
    std::map<std::string, std::string> header;
    std::string data;
    if (!read_record(&header, &data)) break;
    if (header.find("op") == header.end()) break;
    uint8_t op = (uint8_t)header["op"][0];
    if (op == BAG_OP_CONNECTION) parse_connection(header, data);
    else if (op == BAG_OP_CHUNK) read_chunk(header, data, cb);
    else if (op == BAG_OP_MSG_DATA) { BagMessage m = msg_from(header, data); cb(m, conns_[m.conn]); }
  }
}

BagReader::Info BagReader::info() {
  Info out;
  out.path = path_;
  // connections (file-level) + chunk infos, from the index tail.
  f_.clear();
  f_.seekg((std::streamoff)index_pos_);
  for (uint32_t k = 0; k < conn_count_; ++k) {
    std::map<std::string, std::string> header;
    std::string data;
    if (!read_record(&header, &data)) break;
    if ((uint8_t)header["op"][0] == BAG_OP_CONNECTION) parse_connection(header, data);
  }
  bool have_time = false;
  f_.clear();
  f_.seekg((std::streamoff)index_pos_);
  while (true) {
    std::map<std::string, std::string> header;
    std::string data;
    if (!read_record(&header, &data)) break;
    if (header.find("op") == header.end()) break;
    if ((uint8_t)header["op"][0] != BAG_OP_CHUNK_INFO) continue;
    out.chunks++;
    double s = get_u32(header["start_time"], 0) + get_u32(header["start_time"], 4) * 1e-9;
    double e = get_u32(header["end_time"], 0) + get_u32(header["end_time"], 4) * 1e-9;
    if (!have_time || s < out.start) out.start = s;
    if (!have_time || e > out.end) out.end = e;
    have_time = true;
    for (size_t j = 0; j + 8 <= data.size(); j += 8) {
      int cid = (int)get_u32(data, j);
      uint32_t n = get_u32(data, j + 4);
      out.counts[cid] += n;
      out.messages += n;
    }
  }
  out.conns = conns_;
  std::ifstream sz(path_, std::ios::binary | std::ios::ate);
  out.size = (uint64_t)sz.tellg();
  return out;
}

}  // namespace irap_noroslib

File: python/irap_noroslib/bag.py
"""Read and write ROS bag files (format v2.0) -- no ROS installed.

A bag is ROS's on-disk container for recorded messages. This implements the
real, documented v2.0 layout (http://wiki.ros.org/Bags/Format/2.0), so a bag
written here opens in genuine `rosbag info` / `rosbag play` / `rqt_bag`, and a
bag recorded by real ROS reads back here.

    with BagWriter("out.bag") as w:
        conn = w.connection("/chatter", "std_msgs/String", md5, definition)
        w.write(conn, secs, nsecs, body_bytes)

    for topic, conn, t, body in BagReader("out.bag").read_messages():
        ...

Only the uncompressed encoding is written (compression="none"); a compressed
bag written elsewhere still reads, chunk by chunk, if bz2/lz4 is available.
"""
import bz2
import struct
import time

# Record op codes (the 'op' header field, one byte).
OP_MSG_DATA = 0x02
OP_BAG_HEADER = 0x03
OP_IDX_DATA = 0x04
OP_CHUNK = 0x05
OP_CHUNK_INFO = 0x06
OP_CONNECTION = 0x07

_CHUNK_THRESHOLD = 768 * 1024      # rosbag's default chunk size


# ------------------------------------------------------------ header codec ----
def _enc_field(name, value_bytes):
    field = name.encode("ascii") + b"=" + value_bytes
    return struct.pack("<I", len(field)) + field


def encode_header_fields(fields):
    """fields: list of (name, value_bytes) -> the raw field sequence, no outer
    length. This is exactly a connection record's DATA payload."""
    return b"".join(_enc_field(n, v) for n, v in fields)


def encode_header(fields):
    """A length-prefixed header block: <4-byte len><fields>. Used for the HEADER
    part of every record (the outer data_len delimits DATA, so DATA payloads that
    happen to be headers -- connections -- use encode_header_fields instead)."""
    body = encode_header_fields(fields)
    return struct.pack("<I", len(body)) + body


def _decode_header_bytes(blob):
    """blob is the header body (no outer length). -> {name: value_bytes}."""
    out = {}
    i = 0
    while i < len(blob):
        (flen,) = struct.unpack_from("<I", blob, i)
        i += 4
        field = blob[i:i + flen]
        i += flen
        eq = field.index(b"=")
        out[field[:eq].decode("ascii")] = field[eq + 1:]
    return out


def _u8(v):
    return struct.pack("<B", v)


def _u32(v):
    return struct.pack("<I", v)


def _u64(v):
    return struct.pack("<Q", v)


def _time(secs, nsecs):
    return struct.pack("<II", secs, nsecs)


# --------------------------------------------------------------- connection ----
class _Conn:
    def __init__(self, conn_id, topic, type_name, md5, definition, callerid="", latching=""):
        self.id = conn_id
        self.topic = topic
        self.type = type_name
        self.md5 = md5
        self.definition = definition
        self.callerid = callerid
        self.latching = latching
        self.count = 0

    def header_data(self):
        """The connection header stored in the record's DATA field."""
        fields = [
            ("topic", self.topic.encode("utf-8")),
            ("type", self.type.encode("utf-8")),
            ("md5sum", self.md5.encode("ascii")),
            ("message_definition", self.definition.encode("utf-8")),
        ]
        if self.callerid:
            fields.append(("callerid", self.callerid.encode("utf-8")))
        if self.latching:
            fields.append(("latching", self.latching.encode("ascii")))
        return encode_header_fields(fields)


# ------------------------------------------------------------------ writer ----
class BagWriter:
    """Write a valid, real-ROS-readable v2.0 bag."""

    def __init__(self, path):
        self.path = path
        self.f = open(path, "wb")
        self.f.write(b"#ROSBAG V2.0\n")
        # Reserve the bag-header record; rewritten at close with real offsets.
        self._bagheader_pos = self.f.tell()
        self.f.write(self._bag_header_record(0, 0, 0))
        self._conns = {}                 # topic -> _Conn
        self._conn_by_id = {}
        self._chunk_buf = []             # (conn_id, secs, nsecs, body)
        self._chunk_bytes = 0
        self._chunk_infos = []           # (chunk_pos, start, end, {conn: count})
        self._start_time = None
        self._end_time = None

    # -- public -----------------------------------------------------------
    def connection(self, topic, type_name, md5, definition, callerid="", latching=""):
        c = self._conns.get(topic)
        if c is None:
            cid = len(self._conns)
            c = _Conn(cid, topic, type_name, md5, definition, callerid, latching)
            self._conns[topic] = c
            self._conn_by_id[cid] = c
        return c

    def write(self, conn, secs, nsecs, body):
        self._chunk_buf.append((conn.id, secs, nsecs, body))
        self._chunk_bytes += len(body)
        conn.count += 1
        t = (secs, nsecs)
        if self._start_time is None or t < self._start_time:
            self._start_time = t
        if self._end_time is None or t > self._end_time:
            self._end_time = t
        if self._chunk_bytes >= _CHUNK_THRESHOLD:
            self._flush_chunk()

    def close(self):
        if self.f is None:
            return
        self._flush_chunk()
        index_pos = self.f.tell()
        # File-level connection records, then chunk-info records.
        for c in self._conn_by_id.values():
            self._write_record(
                [("op", _u8(OP_CONNECTION)), ("conn", _u32(c.id)),
                 ("topic", c.topic.encode("utf-8"))], c.header_data())
        for chunk_pos, start, end, counts in self._chunk_infos:
            data = b"".join(_u32(cid) + _u32(n) for cid, n in counts.items())
            self._write_record(
                [("op", _u8(OP_CHUNK_INFO)), ("ver", _u32(1)),
                 ("chunk_pos", _u64(chunk_pos)),
                 ("start_time", _time(*start)), ("end_time", _time(*end)),
                 ("count", _u32(len(counts)))], data)
        # Rewrite the bag header now that we know the offsets.
        self.f.seek(self._bagheader_pos)
        self.f.write(self._bag_header_record(index_pos, len(self._conns),
                                             len(self._chunk_infos)))
        self.f.close()
        self.f = None

    def __enter__(self):
        return self

    def __exit__(self, *a):
        self.close()

    # -- internals --------------------------------------------------------
    def _write_record(self, header_fields, data):
        self.f.write(encode_header(header_fields))
        self.f.write(_u32(len(data)))
        self.f.write(data)

    def _bag_header_record(self, index_pos, conn_count, chunk_count):
        header = encode_header([
            ("op", _u8(OP_BAG_HEADER)),
            ("index_pos", _u64(index_pos)),
            ("conn_count", _u32(conn_count)),
            ("chunk_count", _u32(chunk_count)),
        ])
        # rosbag pads the bag-header record's DATA so the whole record is >=4096
        # bytes, leaving room to rewrite the counts in place.
        pad_len = 4096 - len(header) - 4
        if pad_len < 0:
            pad_len = 0
        return header + _u32(pad_len) + (b" " * pad_len)

    def _flush_chunk(self):
        if not self._chunk_buf:
            return
        chunk_pos = self.f.tell()
        # Build the uncompressed chunk data: connection records (first sight in
        # this chunk) interleaved with message-data records; record per-conn
        # offsets for the index.
        data = bytearray()
        seen = set()
        index = {}                       # conn_id -> [(secs, nsecs, offset)]
        counts = {}
        for cid, secs, nsecs, body in self._chunk_buf:
            c = self._conn_by_id[cid]
            if cid not in seen:
                seen.add(cid)
                data += self._record_bytes(
                    [("op", _u8(OP_CONNECTION)), ("conn", _u32(cid)),
                     ("topic", c.topic.encode("utf-8"))], c.header_data())
            offset = len(data)
            data += self._record_bytes(
                [("op", _u8(OP_MSG_DATA)), ("conn", _u32(cid)),
                 ("time", _time(secs, nsecs))], body)
            index.setdefault(cid, []).append((secs, nsecs, offset))
            counts[cid] = counts.get(cid, 0) + 1
        # Chunk record: compression=none, size=uncompressed length.
        self._write_record(
            [("op", _u8(OP_CHUNK)), ("compression", b"none"),
             ("size", _u32(len(data)))], bytes(data))
        # Index records: one per connection in this chunk.
        for cid, entries in index.items():
            idx = b"".join(_time(s, n) + _u32(off) for s, n, off in entries)
            self._write_record(
                [("op", _u8(OP_IDX_DATA)), ("ver", _u32(1)),
                 ("conn", _u32(cid)), ("count", _u32(len(entries)))], idx)
        # Remember chunk info for the file tail.
        starts = min((s, n) for cid, s, n, _o in
                     ((cid, s, n, o) for cid in index for (s, n, o) in index[cid]))
        ends = max((s, n) for cid in index for (s, n, o) in index[cid])
        self._chunk_infos.append((chunk_pos, starts, ends, counts))
        self._chunk_buf = []
        self._chunk_bytes = 0

    @staticmethod
    def _record_bytes(header_fields, data):
        return encode_header(header_fields) + _u32(len(data)) + data


# ------------------------------------------------------------------ reader ----
class BagMessage:
    __slots__ = ("topic", "conn", "secs", "nsecs", "data")

    def __init__(self, topic, conn, secs, nsecs, data):
        self.topic = topic
        self.conn = conn
        self.secs = secs
        self.nsecs = nsecs
        self.data = data

    @property
    def t(self):
        return self.secs + self.nsecs * 1e-9


class BagReader:
    """Read a v2.0 bag: iterate messages, or summarise it."""

    def __init__(self, path):
        self.path = path
        self.f = open(path, "rb")
        line = self.f.readline()
        if not line.startswith(b"#ROSBAG V2.0"):
            raise ValueError("%s is not a ROS bag v2.0 file" % path)
        self._conns = {}                 # id -> dict(topic,type,md5,definition)
        self._read_bag_header()

    def _read_record(self):
        raw = self.f.read(4)
        if len(raw) < 4:
            return None, None
        (hlen,) = struct.unpack("<I", raw)
        header = _decode_header_bytes(self.f.read(hlen))
        (dlen,) = struct.unpack("<I", self.f.read(4))
        data = self.f.read(dlen)
        return header, data

    def _read_bag_header(self):
        pos = self.f.tell()
        header, _data = self._read_record()
        self._conn_count = struct.unpack("<I", header["conn_count"])[0]
        self._chunk_count = struct.unpack("<I", header["chunk_count"])[0]
        self._index_pos = struct.unpack("<Q", header["index_pos"])[0]
        self._data_start = self.f.tell()
        self.f.seek(pos)                 # leave positioned at bag header

    def _parse_connection(self, header, data):
        cid = struct.unpack("<I", header["conn"])[0]
        fields = _decode_header_bytes(data)
        self._conns[cid] = {
            "topic": header.get("topic", fields.get("topic", b"")).decode("utf-8"),
            "type": fields.get("type", b"").decode("utf-8"),
            "md5": fields.get("md5sum", b"").decode("ascii"),
            "definition": fields.get("message_definition", b"").decode("utf-8", "replace"),
            "callerid": fields.get("callerid", b"").decode("utf-8", "replace"),
        }
        return cid

    def read_messages(self):
        """Yield BagMessage in file order (chunk by chunk, message by message)."""
        self.f.seek(self._data_start)
        while True:
            header, data = self._read_record()
            if header is None:
                break
            op = header["op"][0]
            if op == OP_CONNECTION:
                self._parse_connection(header, data)
            elif op == OP_CHUNK:
                yield from self._read_chunk(header, data)
            elif op in (OP_IDX_DATA, OP_CHUNK_INFO):
                pass                      # tail index; messages already yielded
            elif op == OP_MSG_DATA:       # (bags without chunks are legal)
                yield self._msg_from(header, data)
            # once we reach the file-level connection/index tail we could stop,
            # but reading to EOF is simplest and robust.

    def _read_chunk(self, header, data):
        comp = header.get("compression", b"none").decode("ascii")
        if comp == "bz2":
            data = bz2.decompress(data)
        elif comp == "lz4":
            try:
                import lz4.frame
                data = lz4.frame.decompress(data)
            except Exception as e:  # noqa
                raise ValueError("lz4-compressed bag needs the lz4 package: %s" % e)
        elif comp != "none":
            raise ValueError("unknown chunk compression %r" % comp)
        i = 0
        while i < len(data):
            (hlen,) = struct.unpack_from("<I", data, i)
            i += 4
            rhdr = _decode_header_bytes(data[i:i + hlen])
            i += hlen
            (dlen,) = struct.unpack_from("<I", data, i)
            i += 4
            rdata = data[i:i + dlen]
            i += dlen
            op = rhdr["op"][0]
            if op == OP_CONNECTION:
                self._parse_connection(rhdr, rdata)
            elif op == OP_MSG_DATA:
                yield self._msg_from(rhdr, rdata)

    def _msg_from(self, header, data):
        cid = struct.unpack("<I", header["conn"])[0]
        secs, nsecs = struct.unpack("<II", header["time"])
        topic = self._conns.get(cid, {}).get("topic", "?")
        return BagMessage(topic, cid, secs, nsecs, data)

    def connections(self):
        """Read the file-level connection records (fast; no chunk scan)."""
        if not self._conns:
            self.f.seek(self._index_pos)
            for _ in range(self._conn_count):
                header, data = self._read_record()
                if header is None:
                    break
                if header["op"][0] == OP_CONNECTION:
                    self._parse_connection(header, data)
        return self._conns

    def chunk_infos(self):
        """(chunk_pos, start, end, {conn: count}) for each chunk, from the tail."""
        self.connections()
        infos = []
        self.f.seek(self._index_pos)
        while True:
            header, data = self._read_record()
            if header is None:
                break
            if header["op"][0] != OP_CHUNK_INFO:
                continue
            start = struct.unpack("<II", header["start_time"])
            end = struct.unpack("<II", header["end_time"])
            chunk_pos = struct.unpack("<Q", header["chunk_pos"])[0]
            counts = {}
            for j in range(0, len(data), 8):
                cid, n = struct.unpack_from("<II", data, j)
                counts[cid] = n
            infos.append((chunk_pos, start, end, counts))
        return infos

    def close(self):
        if self.f:
            self.f.close()
            self.f = None
